### PR TITLE
Dawn GPU backends: Bug in code generation

### PR DESCRIPTION
**Pull request by stubbiali**
## Description

In the `backend_pre_run` method of `DawnCUDAPyModuleGenerator`, the string within the list comprehension is converted to an f-string. Plus, along the lines of https://github.com/GridTools/gt4py/pull/217, a safety check is added to disregard unused optional fields.

## Requirements

Before submitting this PR, please make sure:

- [ ] The code builds cleanly without new errors or warnings
- [ ] The code passes all the existing tests
- [ ] If this PR adds a new feature, new tests have been added to test these
new features
- [ ] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [ ] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [ ] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

File: tests/test_unittest/test_backend.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 from gt4py.gtscript import __INLINED, PARALLEL, Field, computation, interval
 from gt4py.stencil_builder import StencilBuilder
 
-from ..definitions import ALL_BACKENDS
+from ..definitions import ALL_BACKENDS, CPU_BACKENDS, DAWN_CPU_BACKENDS
```
Comment (jdahm) [inline tests/test_unittest/test_backend.py:9]: Should `DAWN_CPU_BACKENDS` go in `CPU_BACKENDS`?

File: tests/test_unittest/test_backend.py [hunk 2]
```diff
@@ -83,8 +83,9 @@ def test_generate_pre_run(backend_name, mode):
     module_generator.args_data = args_data
     source = module_generator.generate_pre_run()
 
-    if backend_cls.storage_info["device"] == "cpu":
-        assert source == ""
+    if backend_name in CPU_BACKENDS:
+        if backend_name not in DAWN_CPU_BACKENDS:
+            assert source == ""
```
Comment (jdahm) [inline tests/test_unittest/test_backend.py:88]: This assertion could be simplified to `assert source or backend_name in DAWN_CPU_BACKENDS`

File: src/gt4py/backend/dawn_backends.py [hunk 3]
```diff
@@ -401,8 +401,12 @@ def backend_imports(self) -> List[str]:
         return ["import cupy"]
 
     def backend_pre_run(self) -> List[str]:
-        field_names = self.args_data["field_info"].keys()
-        return ["{name}.host_to_device()" for name in field_names]
+        field_names = [
+            key
+            for key in self.args_data["field_info"]
+            if self.args_data["field_info"][key] is not None
+        ]
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:408]: I think this is slighty more readable:
```suggestion
        field_names = [
            key
            for key, value in self.args_data["field_info"].items()
            if value is not None
        ]
```

File: tests/test_unittest/test_backend.py (hunk 2 above)
Comment (egparedes) [reply tests/test_unittest/test_backend.py:88]: :+1:

**Review by egparedes — COMMENTED** [review]
Thanks for submitting a bug fix. It's almost ready to be merged. Just a couple of minor comments to increase the readability of the code

File: tests/test_unittest/test_backend.py (hunk 2 above)
Comment (stubbiali) [reply tests/test_unittest/test_backend.py:88]: I know it's not pretty, but I formulated the code in this way so to make clear that we need to diversify between GT and Dawn backends.

Comment (stubbiali) [reply tests/test_unittest/test_backend.py:88]: Plus it may serve as a remainder that Dawn backends need more meticulous testing here.

**Review by egparedes — APPROVED** [review]
Ready to merge.

**Pull request merged**